crates/ruff_python_formatter/src/expression/expr_slice.rs

```diff
@@ -1,23 +1,257 @@
-            f,
-            [not_yet_implemented_custom_text(
-                "NOT_IMPLEMENTED_start:NOT_IMPLEMENTED_end"
-            )]
-        )
+    /// This implementation deviates from black in that comments are attached to the section of the
+    /// slice they originate in
+    fn fmt_fields(&self, item: &ExprSlice, f: &mut PyFormatter) -> FormatResult<()> {
+        // `[lower:upper:step]`
+        let ExprSlice {
+            range,
+            lower,
+            upper,
+            step,
+        } = item;
+
+        let (first_colon, second_colon) =
+            find_colons(f.context().contents(), *range, lower, upper)?;
+
+        // Handle comment placement
+        // In placements.rs, we marked comment for None nodes a dangling and associated all others
+        // as leading or dangling wrt to a node. That means we either format a node and only have
+        // to handle newlines and spacing, or the node is None and we insert the corresponding
+        // slice of dangling comments
+        let comments = f.context().comments().clone();
+        let slice_dangling_comments = comments.dangling_comments(item.as_any_node_ref());
+        // Put the dangling comments (where the nodes are missing) into buckets
+        let first_colon_partition_index = slice_dangling_comments
+            .partition_point(|x| x.slice().start() < first_colon.range.start());
+        let second_colon_partition_index =
+            second_colon
+                .as_ref()
+                .map_or(slice_dangling_comments.len(), |second_colon| {
+                    slice_dangling_comments
+                        .partition_point(|x| x.slice().start() < second_colon.range.start())
+                });
+        // Ensure there a no dangling comments for a node if the node is present
+        debug_assert!(!(lower.is_some() && first_colon_partition_index > 0));
+        debug_assert!(
+            !(upper.is_some() && first_colon_partition_index < second_colon_partition_index)
+        );
+        debug_assert!(
+            !(step.is_some() && second_colon_partition_index < slice_dangling_comments.len())
+        );
+
+        // Handle spacing around the colon(s)
+        // https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#slices
+        let lower_simple = lower.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let upper_simple = upper.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let step_simple = step.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let all_simple = lower_simple && upper_simple && step_simple;
+
+        // lower
+        if let Some(lower) = lower {
+            lower.format().fmt(f)?;
+            let lower_trailing_comments = comments.trailing_comments(lower.as_ref().into());
+            if !lower_trailing_comments.is_empty() {
+                // make sure the colon is after the comments
+                hard_line_break().fmt(f)?;
+            }
```

Comment:
Would the following also work? It would be cheaper because querying comments is somewhat expensive (not very but it requires a hash map lookup). 

```rust
write!(f, [group(&format_args[lower.format(), soft_line_break(), first_colon_space, text(:)])]?;
```

Or does that not work because we want to keep the `:` on the same line as `lower` even if it expands?

Reply:
that sounds like a really complex rewrite and i'd like to avoid that :grimacing:

Reply:
You could give it a try for a single expression AND/OR try it as a separate PR.